kinto/plugins/quotas/listener.py

```diff
@@ -0,0 +1,212 @@
+    return settings.get(
+        # Bucket specific
+        'quotas.bucket_{}_{}'.format(bucket_id, name),
+        # Global to all buckets
+        settings.get('quotas.bucket_{}'.format(name), None))
+
+
+def get_collection_settings(settings, bucket_id, collection_id, name):
+    return settings.get(
+        # Specific for a given bucket collection
+        'quotas.collection_{}_{}_{}'.format(bucket_id, collection_id, name),
+        # Specific to given bucket collections
+        settings.get('quotas.collection_{}_{}'.format(bucket_id, name),
+                     # Global to all buckets collections
+                     settings.get('quotas.collection_{}'.format(name), None)))
+
+
+def on_resource_changed(event):
+    """
+    Everytime an object is created/changed/deleted, we update the
+    bucket counters. The entries are served as read-only in the
+    :mod:`kinto.plugins.quotas.views` module.
+
+    If a new object override the quotas, we reject the request.
+    """
+    payload = copy.deepcopy(event.payload)
+    action = payload['action']
+    resource_name = payload['resource_name']
+    event_uri = payload['uri']
+
+    settings = event.request.registry.settings
+
+    bucket_id = payload.pop('bucket_id')
+    bucket_uri = instance_uri(event.request, 'bucket', id=bucket_id)
+    collection_id = None
+    collection_uri = None
+    if 'collection_id' in payload:
+        collection_id = payload['collection_id']
+        collection_uri = instance_uri(event.request,
+                                      'collection',
+                                      bucket_id=bucket_id,
+                                      id=collection_id)
+
+    bucket_max_bytes = get_bucket_settings(settings, bucket_id, 'max_bytes')
+    bucket_max_items = get_bucket_settings(settings, bucket_id, 'max_items')
+    bucket_max_bytes_per_item = get_bucket_settings(settings, bucket_id,
+                                                    'max_bytes_per_item')
+    collection_max_bytes = get_collection_settings(settings, bucket_id,
+                                                   collection_id, 'max_bytes')
+    collection_max_items = get_collection_settings(settings, bucket_id,
+                                                   collection_id, 'max_items')
+    collection_max_bytes_per_item = get_collection_settings(
+        settings, bucket_id, collection_id, 'max_bytes_per_item')
+
+    # XXX: Maybe we want to differenciate between
+    #      bucket/collection/records/group payload.
+    max_bytes_per_item = (collection_max_bytes_per_item or
+                          bucket_max_bytes_per_item)
+
+    storage = event.request.registry.storage
+
+    if action == 'delete' and resource_name == 'bucket':
+        try:
+            storage.delete(parent_id=bucket_uri,
+                           collection_id='quota',
+                           object_id='bucket_info')
+        except RecordNotFoundError:  # Pragma: no cover
+            pass
```

Comment:
Actually it can happen if the collection is deleted and the plugin wasn't activated before the last change.